Sources/Controllers/Map/Widgets/Widgets/GlideAverageWidget.swift

```diff
@@ -15,30 +15,59 @@ final class GlideAverageWidget: GlideBaseWidget {
     static let measuredIntervalPrefID = "average_glide_measured_interval_millis"
     private static var availableIntervals: [Int: String] = getAvailableIntervals()
 
+    private let widgetState: GlideAverageWidgetState?
     private let averageGlideComputer = AverageGlideComputer.shared
     private var measuredIntervalPref: OACommonLong
     private var cachedFormattedGlideRatio: String?
+    private var forceUpdate = false // Becomes 'true' when widget state switches
 
-    init(customId: String?, appMode: OAApplicationMode, widgetParams: ([String: Any])? = nil) {
+    init(with widgetState: GlideAverageWidgetState, customId: String?, appMode: OAApplicationMode, widgetParams: ([String: Any])? = nil) {
+        self.widgetState = widgetState
         measuredIntervalPref = Self.registerMeasuredIntervalPref(customId)
         super.init(WidgetType.glideAverage, customId: customId, appMode: appMode, widgetParams: widgetParams)
         updateInfo()
-        setIcon("widget_glide_ratio_average")
+        onClickFunction = { [weak self] _ in
+            guard let self else { return }
+
+            forceUpdate = true
+            self.widgetState?.changeToNextState()
+            updateInfo()
+            setContentTitle(getWidgetName())
+            setIcon(getWidgetIcon())
+        }
+        setContentTitle(getWidgetName())
+        setIcon(getWidgetIcon())
     }
 
     override init(frame: CGRect) {
         measuredIntervalPref = Self.registerMeasuredIntervalPref(nil)
+        widgetState = GlideAverageWidgetState(nil)
         super.init(frame: frame)
     }
 
     required init?(coder: NSCoder) {
         fatalError("init(coder:) has not been implemented")
     }
 
+    override func getWidgetState() -> OAWidgetState? {
+        return widgetState
+    }
+
     override func getSettingsData(_ appMode: OAApplicationMode) -> OATableDataModel? {
         let data = OATableDataModel()
         let section = data.createNewSection()
-        section.headerText = localizedString("average_glide_ratio_time_interval_desc")
+        section.headerText = localizedString("shared_string_settings")
+
+        if let preference = widgetState?.getPreference() {
+            let settingRow = section.createNewRow()
+            settingRow.cellType = OAValueTableViewCell.reuseIdentifier
+            settingRow.key = "value_pref"
+            settingRow.title = localizedString("shared_string_mode")
+            settingRow.setObj(preference, forKey: "pref")
+            settingRow.setObj(getWidgetName() ?? localizedString("average_glide_ratio"), forKey: "value")
```

Comment:
When can getWidgetName be nil?

Reply:
When widgetState is nil: If the widgetState is nil, the method attempts to return widgetType?.title. If widgetType itself is nil or if it is non-nil but its title property is not set, then the method will return nil.